plugin/trino-pulsar/src/test/java/io/trino/plugin/pulsar/TestPulsarRecordCursor.java

```diff
@@ -0,0 +1,496 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.pulsar;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.airlift.log.Logger;
+import io.trino.plugin.pulsar.mock.MockCursor;
+import io.trino.plugin.pulsar.mock.MockManagedLedgerFactory;
+import io.trino.spi.predicate.TupleDomain;
+import io.trino.spi.type.RowType;
+import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.client.api.schema.SchemaInfoProvider;
+import org.apache.pulsar.client.impl.schema.KeyValueSchema;
+import org.apache.pulsar.shade.io.netty.buffer.Unpooled;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.AsyncCallbacks;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.Entry;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.ManagedLedgerConfig;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.ManagedLedgerException;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.ManagedLedgerFactory;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.Position;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.ReadOnlyCursor;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.impl.EntryImpl;
+import org.apache.pulsar.shade.org.apache.bookkeeper.mledger.impl.PositionImpl;
+import org.apache.pulsar.shade.org.apache.bookkeeper.stats.NullStatsProvider;
+import org.apache.pulsar.shade.org.apache.pulsar.common.naming.TopicName;
+import org.apache.pulsar.shade.org.apache.pulsar.common.schema.KeyValue;
+import org.apache.pulsar.shade.org.apache.pulsar.common.schema.KeyValueEncodingType;
+import org.apache.pulsar.shade.org.apache.pulsar.common.schema.SchemaInfo;
+import org.testng.annotations.Test;
+
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.Function;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertTrue;
+
+@Test(singleThreaded = true)
+public class TestPulsarRecordCursor
```

Comment:
I don't really care about such tests. They are hard to maintain and do not help much. 

I would suggest to you implement `BaseConnectorTest` for pulsar. Please also see other connectors to see how you can test type mapping. We prefer to run entire query and see the results that user could see. That way functionality is covered and even if you change internals of product code, the tests are unchanged and they are up to date providing the coverage.

That is the reason why don't like excessive mocking, not that we don't like mockito.

Reply:
Added a test based on `BaseConnectorTest`, disabled some test not applicable to Pulsar connector for now like create/drop table, or describe table as Pulsar table has some internal columns.